src/Azure/Azure.Quantum.Client/Machine/QuantumMachineFactory.cs

```diff
@@ -17,21 +17,26 @@ public static class QuantumMachineFactory
         /// <returns>A quantum machine for job submission targeting <c>targetName</c>.</returns>
         public static IQuantumMachine? CreateMachine(Workspace workspace, string targetName, string storageAccountConnectionString)
         {
-            if (string.IsNullOrEmpty(targetName))
-            {
-                return null;
-            }
-
             var machineName =
-                targetName.StartsWith("ionq.")
+                targetName is null
+                ? null
+                : targetName.StartsWith("ionq.")
                 ? "Microsoft.Quantum.Providers.IonQ.Targets.IonQQuantumMachine, Microsoft.Quantum.Providers.IonQ"
                 : targetName.StartsWith("honeywell.")
                 ? "Microsoft.Quantum.Providers.Honeywell.Targets.HoneywellQuantumMachine, Microsoft.Quantum.Providers.Honeywell"
                 : null;
-            return machineName is null
+
+            // First try to load the signed assembly with the correct version, then try the unsigned one.
+            var machineType =
+                machineName is null
+                ? null
+                : Type.GetType($"{machineName}, Version={typeof(Workspace).Assembly.GetName().Version}, Culture=neutral, PublicKeyToken=40866b40fd95c7f5")
```

Comment:
I checked several different signed versions from our CI build of both the IonQ and Honeywell assemblies and they all had this same `PublicKeyToken` value.

Reply:
The version seems to be required, but it seems safe to just enforce that it matches the currently-loaded version of `Microsoft.Azure.Quantum.Client`.

Reply:
I get a different error message now:

```
Unhandled exception: System.IO.FileLoadException: Could not load file or assembly 'Microsoft.Quantum.Providers.IonQ, Version=1.0.0.0, Culture=neutral, PublicKeyToken=40866b40fd95c7f5'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)
File name: 'Microsoft.Quantum.Providers.IonQ, Version=1.0.0.0, Culture=neutral, PublicKeyToken=40866b40fd95c7f5'
   at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, StackCrawlMarkHandle stackMark, ObjectHandleOnStack assemblyLoadContext, Boolean loadTypeFromPartialName, ObjectHandleOnStack type, ObjectHandleOnStack keepalive)
   at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, StackCrawlMark& stackMark, AssemblyLoadContext assemblyLoadContext, Boolean loadTypeFromPartialName)
   at System.RuntimeType.GetType(String typeName, Boolean throwOnError, Boolean ignoreCase, StackCrawlMark& stackMark)
   at System.Type.GetType(String typeName)
   at Microsoft.Azure.Quantum.QuantumMachineFactory.CreateMachine(IWorkspace workspace, String targetName, String storageAccountConnectionString) in C:\Users\samarsha\Source\Repos\microsoft\qsharp-runtime\src\Azure\Azure.Quantum.Client\Machine\QuantumMachineFactory.cs:line 30
   at Microsoft.Quantum.CsharpGeneration.EntryPointDriver.Azure.CreateMachine(AzureSettings settings) in C:\Users\samarsha\Source\Repos\microsoft\qsharp-runtime\src\Simulation\EntryPointDriver\Azure.cs:line 60
   at Microsoft.Quantum.CsharpGeneration.EntryPointDriver.Azure.Submit[TIn,TOut](IEntryPoint`2 entryPoint, ParseResult parseResult, AzureSettings settings) in C:\Users\samarsha\Source\Repos\microsoft\qsharp-runtime\src\Simulation\EntryPointDriver\Azure.cs:line 28
   at Microsoft.Quantum.CsharpGeneration.EntryPointDriver.Driver`3.Submit(ParseResult parseResult, AzureSettings settings) in C:\Users\samarsha\Source\Repos\microsoft\qsharp-runtime\src\Simulation\EntryPointDriver\Driver.cs:line 124
   at System.CommandLine.Invocation.CommandHandler.GetResultCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseErrorReporting>b__20_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass15_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass24_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__21_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseDirective>b__19_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseDebugDirective>b__11_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass13_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

I think this is because I'm using a locally built version of Microsoft.Azure.Quantum.Client (version 1.0.0.0), but a a packaged version of IonQ with a different version number. My local setup is hacky, but it was the only way I was able to get something working where I could test local changes to the entry point driver with the quantum machines (without making a full Solid build to synchronize the version numbers).

It looks like `Type.GetType` is throwing this error even though the `throwOnError` parameter is not set to true. It's not ideal, but maybe the exception needs to be caught so the next call to `Type.GetType` can run?

Reply:
Yes, that sounds like the right approach. I shouldn't have assumed that `throwOnError=false` means that it won't throw on error.

Reply:
Or, wait, actually, since you're using a project reference to Microsoft.Azure.Quantum.Client, you'll probably always be using a locally-built version and will never have the correct version number.

Could you use a package reference to Microsoft.Azure.Quantum.Client instead? I'm using version 0.11.2005.1420-beta and things are working.

Reply:
Well, I seem to have broken my local project after changing it a bit and can't get it back to the state it was in before. I'll approve the PR since the new change fixes the assembly loading problem - I think the other problem I'm running into is unrelated. :)

About using a package reference, that might work, but since Microsoft.Azure.Quantum.Client is in the same repository as EntryPointDriver I'm not sure if it makes sense - I think all of the projects I have seen use project references to other projects in the same repository, and package references to projects in different repositories.

My idea now to avoid version conflicts when testing locally was to just build the entire Solid repository locally, so everything would have the same version, but at the moment I'm having some issues with that. :) Do you want to remove the try/catch and require that the versions match?

Reply:
I think I'll leave it this way for now, since at least it should allow you to continue with your existing workflow (if you can get that working again). If/when you figure out a way to get around the version issue, you could always remove the try/catch at that time. I don't think it hurts anything to have it there, though.

Thanks for reviewing and working through this with me!